gr-digital/lib/vblast_decoder_cc_impl.cc

```diff
@@ -0,0 +1,237 @@
+            Eigen::MatrixXcf csi_matrix(d_num_inputs, d_num_inputs);
+            for (int i = 0; i < d_num_inputs; ++i) {
+              csi_matrix.row(i) = Eigen::VectorXcf::Map(&d_csi[i][0], d_csi[i].size());
+            }
+            // Calculate the pseudo-inverse fo the CSI matrix.
+            Eigen::MatrixXcf snr_matrix = Eigen::MatrixXcf::Zero(d_num_inputs, d_num_inputs);
+            for (int j = 0; j < d_num_inputs; ++j) {
+              snr_matrix.coeffRef(j, j) = (gr_complex)(1.0 / d_snr[j]);
+            }
+            Eigen::MatrixXcf csi_pseudo_inverse = (csi_matrix.adjoint()*csi_matrix + snr_matrix).inverse() * csi_matrix.adjoint();
+
+            // Map the inverse of the Eigen MatrixXcf to the 2-dim equalizer std::vector.
+            for (int i = 0; i < d_num_inputs; ++i) {
+              Eigen::VectorXcf::Map(&d_mimo_equalizer[i][0], csi_matrix.row(i).size()) = csi_pseudo_inverse.row(i);
+            }
+#else
+            throw std::runtime_error("Required library Eigen3 for MxM MIMO schemes with M>2 not installed.");
+#endif
+          }
+        }
+      } else{
+        // The selected equalizer type is not existing in this implementation.
+      }
+    }
+
+    void
+    vblast_decoder_cc_impl::equalize_symbol(gr_vector_const_void_star input,
+                                            gr_complex* out,
+                                            uint32_t offset,
+                                            uint32_t length) {
+      std::fill(out, &out[length*d_num_inputs], 0.0);
+      for (int n = 0; n < d_num_inputs; ++n) {
+        gr_complex *in = &((gr_complex *) input[n])[offset/d_num_inputs];
+        for (unsigned int i = 0; i < length; ++i) {
+          for (int j = 0; j < d_num_inputs; ++j) {
+            out[i*d_num_inputs + j] += d_mimo_equalizer[j][n] * in[i];
+          }
+        }
+      }
+    }
+
+    int
+    vblast_decoder_cc_impl::work(int noutput_items,
+        gr_vector_const_void_star &input_items,
+        gr_vector_void_star &output_items)
+    {
+      gr_complex *out = (gr_complex *) output_items[0];
+      uint16_t nprocessed = 0; // Number of read and written items.
+
+      // Collect all tags of the input buffer in the vector 'tags'.
+      get_tags_in_window(tags, 0, 0, noutput_items);
+
+      uint16_t symbol_length; // Number of items in the current symbol.
+
+      if(tags.size() == 0){ // Input buffer includes no tags at all.
+        // Handle all samples in buffer as they belong to the current symbol.
+        symbol_length = noutput_items;
+        equalize_symbol(input_items, out, nprocessed, symbol_length);
+        nprocessed += symbol_length;
+      } else { // Input buffer includes tags.
+        if (tags[0].offset - nitems_read(0) > 0){
+          /* There are items in the input buffer, before the first tag arrives,
+           * which belong to the previous symbol. */
+          symbol_length = (tags[0].offset - nitems_read(0))*d_num_inputs;
```

Comment:
or rather, limited to (2¹⁶-1)/num_inputs? Sounds a bit of a dangerous thing, considering let's say 8×8 DAB MIMO ;)

Reply:
I am going to fix these issues in the according branch 'vblast'.